exchanges/gateio/gateio.go

```diff
@@ -0,0 +1,379 @@
+	if err != nil {
+		return res, err
+	}
+	return res, nil
+}
+
+// GetTickers returns tickers for all symbols
+func (g *Gateio) GetTickers() (map[string]TickerResponse, error) {
+	url := fmt.Sprintf("%s/%s/%s", gateioMarketURL, gateioAPIVersion, gateioTickers)
+
+	resp := make(map[string]TickerResponse)
+	err := g.SendHTTPRequest(url, &resp)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// GetOrderbook returns the orderbook data for a suppled symbol
+func (g *Gateio) GetOrderbook(symbol string) (Orderbook, error) {
+	url := fmt.Sprintf("%s/%s/%s/%s", gateioMarketURL, gateioAPIVersion, gateioOrderbook, symbol)
+
+	var resp OrderbookResponse
+	err := g.SendHTTPRequest(url, &resp)
+	if err != nil {
+		return Orderbook{}, err
+	}
+
+	if resp.Result != "true" {
+		return Orderbook{}, errors.New("result was not true")
+	}
+
+	var ob Orderbook
+
+	// Asks are in reverse order
+	for x := len(resp.Asks) - 1; x != 0; x-- {
+		data := resp.Asks[x]
+
+		price, err := strconv.ParseFloat(data[0], 64)
+		if err != nil {
+			continue
+		}
+
+		amount, err := strconv.ParseFloat(data[1], 64)
+		if err != nil {
+			continue
+		}
+
+		ob.Asks = append(ob.Asks, OrderbookItem{Price: price, Amount: amount})
+	}
+
+	for x := range resp.Bids {
+		data := resp.Bids[x]
+
+		price, err := strconv.ParseFloat(data[0], 64)
+		if err != nil {
+			continue
+		}
+
+		amount, err := strconv.ParseFloat(data[1], 64)
+		if err != nil {
+			continue
+		}
+
+		ob.Bids = append(ob.Bids, OrderbookItem{Price: price, Amount: amount})
+	}
+
+	ob.Result = resp.Result
+	ob.Elapsed = resp.Elapsed
+	return ob, nil
+}
+
+// GetSpotKline 返回市场最近时间段内的K先数据
+func (g *Gateio) GetSpotKline(arg KlinesRequestParams) ([]*KLineResponse, error) {
+
+	url := fmt.Sprintf("%s/%s/%s/%s?group_sec=%d&range_hour=%d", gateioMarketURL, gateioAPIVersion, gateioKline, arg.Symbol, arg.GroupSec, arg.HourSize)
+
+	var rawKlines map[string]interface{}
+	err := g.SendHTTPRequest(url, &rawKlines)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []*KLineResponse
+
+	if rawKlines == nil || rawKlines["data"] == nil {
+		return nil, fmt.Errorf("rawKlines is nil. Err: %s", err)
+	}
+
+	//对于 Data数据，再次解析
+	rawKlineDatasString, _ := json.Marshal(rawKlines["data"].([]interface{}))
+	rawKlineDatas := [][]interface{}{}
+	if err := json.Unmarshal(rawKlineDatasString, &rawKlineDatas); err != nil {
+		return nil, fmt.Errorf("rawKlines unmarshal failed. Err: %s", err)
+	}
+	for _, k := range rawKlineDatas {
+		otString, _ := strconv.ParseFloat(k[0].(string), 64)
+		ot, err := common.TimeFromUnixTimestampFloat(otString)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse Kline.OpenTime. Err: %s", err)
+		}
+		_vol, err := common.FloatFromString(k[1])
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse Kline.Volume. Err: %s", err)
+		}
+		_id, err := common.FloatFromString(k[0])
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse Kline.Id. Err: %s", err)
```

Comment:
Just for clarity also here I clearly see different tab widths when selecting the text.